File: src/dp/knapsack/km46.py
# 01 背包

itemNum, totalVolume = map(int, input().strip().split(" "))

itemVolumes = list(map(int, input().strip().split(" ")))
itemValues = list(map(int, input().strip().split(" ")))

f = [[0 for _ in range(totalVolume + 1)] for _ in range(itemNum + 1)]

for i in range(1, itemNum + 1):
    for j in range(1, totalVolume + 1):
        if j < itemVolumes[i - 1]:
            f[i][j] = f[i - 1][j]
            continue

        f[i][j] = max(f[i - 1][j], f[i - 1][j - itemVolumes[i - 1]] + itemValues[i - 1])

print(f[itemNum][totalVolume])

File: src/dp/knapsack/km46.cc
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
  int itemNum, totalVolume;
  cin >> itemNum >> totalVolume;

  auto itemVolumes = vector<int>(itemNum, 0);
  auto itemValues = vector<int>(itemNum, 0);

  for (auto i = 0; i < itemNum; i++) {
    cin >> itemVolumes[i];
  }

  for (auto i = 0; i < itemNum; i++) {
    cin >> itemValues[i];
  }

  auto f = vector<vector<int>>(itemNum + 1, vector<int>(totalVolume + 1, 0));

  for (auto i = 1; i <= itemNum; i++) {
    for (auto j = 1; j <= totalVolume; j++) {
      if (j < itemVolumes[i - 1]) {
        f[i][j] = f[i - 1][j];
        continue;
      }

      f[i][j] = max(f[i - 1][j],
                    f[i - 1][j - itemVolumes[i - 1]] + itemValues[i - 1]);
    }
  }

  cout << f[itemNum][totalVolume] << endl;
}
